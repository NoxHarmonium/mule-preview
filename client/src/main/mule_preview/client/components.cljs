(ns mule-preview.client.components
  "The react components that render the Mule preview"
  (:require
   [reagent.core :as r]
   [clojure.string :refer [split]]
   [mule-preview.client.mappings :refer [element-to-icon-map]]
   [lambdaisland.uri :refer [join]]))

(def default-component-mapping {:image "UnknownNode-48x32.png"})
(def default-category-image "org.mule.tooling.ui.modules.core.miscellaneous.large.png")

(defn- pluralise
  "Pluralise a given string value"
  [string]
  (str string "s"))

(defn- normalise-name [name]
  (let [split-name (split name #":")
        first-component (first split-name)]
    (if (> (count split-name) 1)
      (str first-component ":*")
      first-component)))

(defn- name-to-category-url [name default-image]
  (let [mapping (get element-to-icon-map (keyword name))
        category (:category mapping)
        filename (if (some? category) (str category ".large.png") default-image)]
    (if-not (nil? filename)
      (str "img/icons/" filename)
      nil)))

(defn- name-to-img-url [name is-nested default-value]
  (let [mapping (get element-to-icon-map (keyword name) default-value)
        regular-filename (:image mapping)
        filename (if is-nested (get mapping :nested-image regular-filename) regular-filename)]
    (if-not (nil? filename)
      (str "img/icons/" filename)
      nil)))

(defn- name-to-css-class [name]
  (str "mule-" name))

(defn- image
  ([url content-root] (image url "" content-root))
  ([url class content-root]
   (if-not (nil? url)
     [:img {:src (str (join content-root url)) :class class}]
     nil)))

(defn- child-container [children]
  (into [] (concat [:div {:class "container-children"}] children)))

(defn- arrow [content-root]
  (image "img/arrow-right-2x.png" "flow-arrow" content-root))

(defn mule-component [name description css-class content-root]
  (let [img-url (name-to-img-url name false default-component-mapping)
        category-url (name-to-category-url name default-category-image)]
    [:div {:class ["component" name css-class]}
     (image category-url "category-frame" content-root)
     (image img-url "icon" content-root)
     [:div {:class "label"} description]]))

(defn mule-container [name description children css-class content-root]
  (let [generated-css-class (name-to-css-class name)
        img-url (name-to-img-url name (some? children) nil)
        category-url (name-to-category-url name default-category-image)
        interposed-children (interpose (arrow content-root) children)
        child-container-component (child-container interposed-children)]
    [:div {:class ["container" generated-css-class css-class]}
     [:div {:class "container-title"} description]
     [:div {:class "container-inner"}
      [:div {:class "icon-container"}
       (image category-url "category-frame" content-root)
       (image img-url "icon container-image" content-root)]
      child-container-component]]))