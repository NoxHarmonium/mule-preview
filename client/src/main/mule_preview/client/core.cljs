(ns mule-preview.client.core
  "The entry point to the web application"
  (:require
   [reagent.core :as r]
   [tubax.core :refer [xml->clj]]
   [mule-preview.client.views.preview :refer [start-preview]]
   [mule-preview.client.views.diff :refer [start-diff]]))

(def root-component (r/atom [:div]))

(defn view [root-component]
  [:div {:class "root-component"} @root-component])

(defn mount-root [element root-component]
  (r/render [(partial view root-component)] element))

(defn ^:export mount-diff-on-element [element file-a file-b]
  (let [root-component (r/atom [:div])]
    (start-diff file-a
                file-b
                root-component)
    (mount-root element root-component)))

(defn ^:export mount-preview-on-element [element file]
  (let [root-component (r/atom [:div])]
    (start-preview
     file
     root-component)
    (mount-root element root-component)))

(defn init! []
  (mount-diff-on-element
   (.getElementById js/document "app")
   "/example_xml/nice-example.xml"
   "/example_xml/nice-example-diff.xml"))
