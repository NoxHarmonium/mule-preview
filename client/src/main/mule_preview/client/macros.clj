(ns mule-preview.client.macros
  "Utility functions that transform the code at compile time"
  (:require [cheshire.core :as cc]))

(defmacro get-data [jsonfile]
  "Pulls a JSON file into a data structure at compile time it can be embedded in the final bundle"
  (cc/parse-string (slurp jsonfile) true))

; https://github.com/reagent-project/reagent/wiki/Beware-Event-Handlers-Returning-False
(defmacro handler-fn
  ([& body]
   `(fn [~'event] ~@body nil)))  ;; force return nil