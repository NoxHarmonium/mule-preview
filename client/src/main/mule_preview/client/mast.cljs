(ns mule-preview.client.mast
  "Functions to convert Mule XML data structures to an intermediate MAST structure"
  (:require
   [clojure.walk :refer [prewalk]]
   [mule-preview.client.utils :refer [remove-location]]
   [mule-preview.client.mappings :refer [root-container horizontal-container-list
                                         vertical-container-list error-handler-component-list
                                         error-handler-container-list]]))

(defn- get-tag [node]
  (name (node :tag)))

(defn- is-error-handler [node]
  (let [tag (get-tag node)
        result (contains? error-handler-component-list tag)]
    result))

(defn- get-description [node]
  (let [doc-name (get-in node [:attributes :doc:name])
        name (get-in node [:attributes :name])]
    (or doc-name name)))

(defn- create-mule-component [node tag-name labels]
  (let [description (get-description node)
        content (node :content)
        attributes (:attributes node)]
    {:type :component
     :tag-name tag-name
     :description (or description tag-name)
     :hash (hash (:attributes node))
     :content-hash (hash (remove-location content))
     :labels labels
     :location (:location node)
     :attributes attributes}))

(defn- create-mule-container-component [node tag-name labels]
  (let [description (get-description node)
        content (node :content)
        attributes (:attributes node)]
    {:type :container
     :tag-name tag-name
     :description description
     :hash (hash (:attributes node))
     :content content
     :labels labels
     :location (:location node)
     :attributes attributes}))

(defn- create-mule-psuedo-container [content]
  {:type :container
   :tag-name "psuedo"
   :description ""
   :content content
   :labels #{:horizontal}})

(defn- process-error-container [node tag-name labels]
  (let [description (get-description node)
        content (node :content)
        {error-handlers true regular-components false}
        (group-by is-error-handler content)]
    {:type :error-container
     :tag-name tag-name
     :description description
     :content [(create-mule-psuedo-container regular-components)
               (create-mule-psuedo-container error-handlers)]
     :labels labels}))

(defn- transform-tag [node]
  (let [tag-name (get-tag node)
        is-root-container (= tag-name root-container)
        is-error-handler-container (contains? error-handler-container-list tag-name)
        is-error-handler-component (contains? error-handler-component-list tag-name)
        is-horizontal-container (contains? horizontal-container-list tag-name)
        is-vertical-container (contains? vertical-container-list tag-name)]
    (cond
      is-root-container (create-mule-container-component node tag-name #{:root :vertical})
      is-error-handler-container (process-error-container node tag-name #{:vertical})
      is-error-handler-component (create-mule-container-component node tag-name #{:error-handler})
      is-horizontal-container (create-mule-container-component node tag-name #{:horizontal})
      is-vertical-container (create-mule-container-component node tag-name #{:vertical})
      :else  (create-mule-component node tag-name #{}))))

(defn- transform-fn [node]
  (if (contains? node :tag)
    (let [transformed-tag (transform-tag node)]
      transformed-tag)
    node))

(defn xml->mast [xml]
  (prewalk transform-fn xml))
