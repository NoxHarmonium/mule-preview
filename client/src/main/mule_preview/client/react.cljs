(ns mule-preview.client.react
  "Functions to convert MAST data structures to a structure suitable for reagent (react)"
  (:require
   [clojure.walk :refer [prewalk]]
   [mule-preview.client.mappings :refer [root-container horizontal-container-list
                                         vertical-container-list error-handler-component-list
                                         error-handler-container-list]]
   [mule-preview.client.components :refer [mule-component mule-container]]))

(defn- attributes-to-css [attributes]
  (clojure.string/join " " (map name attributes)))
(defn- create-mule-component [node content-root]
  (let [{:keys [tag-name description attributes location]} node]
    (mule-component {:name tag-name
                     :description description
                     :css-class (attributes-to-css attributes)
                     :content-root content-root
                     :location location})))

(defn- create-mule-container-component [node content-root]
  (let [{:keys [tag-name description content attributes location]} node]
    (mule-container {:name tag-name
                     :description description
                     :children content
                     :css-class (attributes-to-css attributes)
                     :content-root content-root
                     :location location})))

(defn- process-error-container [node content-root]
  (let [{:keys [tag-name description attributes content]} node
        wrapped-content [(create-mule-container-component (first content) content-root)
                         (create-mule-container-component (second content) content-root)]]
    (mule-container {:tag-name tag-name
                     :description description
                     :children wrapped-content
                     :attributes (attributes-to-css attributes)
                     :content-root content-root
                     :location {}})))

(defn- transform-tag [node content-root]
  (let [type (:type node)]
    (case type
      :error-container (process-error-container node content-root)
      :container (create-mule-container-component node content-root)
      :component (create-mule-component node content-root))))

(defn- transform-fn [content-root node]
  (if (contains? node :type)
    (let [transformed-tag (transform-tag node content-root)]
      transformed-tag)
    node))

(defn mast->react [xml content-root]
  (prewalk (partial transform-fn content-root) xml))
