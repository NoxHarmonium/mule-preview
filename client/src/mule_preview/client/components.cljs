(ns mule-preview.client.components
  "The react components that render the Mule preview"
  (:require
   [reagent.core :as r]
   [clojure.string :refer [split]]
   [mule-preview.client.mappings :refer [element-to-icon-map]]))

(def default-component-mapping {:image "UnknownNode-48x32.png"})
(def default-category-image "org.mule.tooling.ui.modules.core.miscellaneous.large.png")

(defn- pluralise
  "Pluralise a given string value"
  [string]
  (str string "s"))

(defn- normalise-name [name]
  (let [split-name (split name #":")
        first-component (first split-name)]
    (if (> (count split-name) 1)
      (str first-component ":*")
      first-component)))

(defn- name-to-category-url [name default-image]
  (let [mapping (get element-to-icon-map (keyword name))
        category (:category mapping)
        filename (if (some? category) (str category ".large.png") default-image)]
    (if-not (nil? filename)
      (str "img/icons/" filename)
      nil)))

(defn- name-to-img-url [name is-nested default-value]
  (let [mapping (get element-to-icon-map (keyword name) default-value)
        regular-filename (:image mapping)
        filename (if is-nested (get mapping :nested-image regular-filename) regular-filename)]
    (if-not (nil? filename)
      (str "img/icons/" filename)
      nil)))

(defn- name-to-css-class [name]
  (str "mule-" name))

(defn- image
  ([url] (image url ""))
  ([url class]
   (if-not (nil? url)
     [:img {:src url :class class}]
     nil)))

(defn- child-container [children]
  (into [] (concat [:div {:class "container-children"}] children)))

(def arrow
  (image "img/arrow-right-2x.png" "flow-arrow"))

(defn mule-component [name description css-class]
  (let [img-url (name-to-img-url name false default-component-mapping)
        category-url (name-to-category-url name default-category-image)]
    [:div {:class ["component" name css-class]}
     (image category-url "category-frame")
     (image img-url "icon")
     [:div {:class "label"} description]]))

(defn mule-container [name description children css-class]
  (let [generated-css-class (name-to-css-class name)
        img-url (name-to-img-url name (some? children) nil)
        category-url (name-to-category-url name default-category-image)
        interposed-children (interpose arrow children)
        child-container-component (child-container interposed-children)]
    [:div {:class ["container" generated-css-class css-class]}
     [:div {:class "container-title"} description]
     [:div {:class "container-inner"}
      [:div {:class "icon-container"}
       (image category-url "category-frame")
       (image img-url "icon container-image")]
      child-container-component]]))