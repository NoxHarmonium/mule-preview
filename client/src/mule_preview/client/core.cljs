(ns mule-preview.client.core
  (:require
   [mule-preview.client.transformer :refer [transform-xml-to-components]]
   [reagent.core :as r]
   [tubax.core :refer [xml->clj]]
   [mule-preview.client.views.preview :refer [start-preview]]
   [mule-preview.client.views.diff :refer [start-diff]]))

(def root-component (r/atom [:div]))

;; -------------------------
;; Views

(defn preview-view []
  (start-preview "/example_xml/nice-example.xml" root-component)
  [:div {:class "root-component"} @root-component])

(defn diff-view []
  (start-diff "/example_xml/nice-example.xml" 
              "/example_xml/nice-example-diff.xml" 
              root-component)
  [:div {:class "root-component"} @root-component])

;; -------------------------
;; Initialize app

(defn mount-root []
  (r/render [preview-view] (.getElementById js/document "app")))

(defn init! []
  (mount-root))
