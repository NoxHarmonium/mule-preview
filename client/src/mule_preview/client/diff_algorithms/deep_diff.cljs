(ns mule-preview.client.diff-algorithms.deep-diff
  "Uses the deep-diff Javscript library to diff data structures.
   Not currently in use since the diffs were not very good to work with.
   See: https://github.com/flitbit/diff")

; Not sure why js->clj doesn't work but this does the trick
; https://stackoverflow.com/a/51439387
(defn clojurise [js-obj]
  (js->clj (-> js-obj js/JSON.stringify js/JSON.parse) :keywordize-keys true))

(defn diff [a b]
  (clojurise (.diff js/DeepDiff (clj->js a) (clj->js b))))