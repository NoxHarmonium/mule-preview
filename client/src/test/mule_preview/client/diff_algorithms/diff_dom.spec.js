import { diff } from "mule-preview/mule_preview.client.diff_algorithms.diff_dom";
import { jsToClj, cljToJs } from "mule-preview/mule_preview.client.test_utils";
import fileAMast from "./__fixtures__/file-a.mast.json";
import fileBMast from "./__fixtures__/file-b.mast.json";

describe("when diffing two MAST data structures", () => {
  it("produces correct output", () => {
    const diffOutput = cljToJs(diff(jsToClj(fileAMast), jsToClj(fileBMast)));
    console.log(JSON.stringify(diffOutput, null, 2));
    expect(diffOutput).toMatchSnapshot();
  });
});
