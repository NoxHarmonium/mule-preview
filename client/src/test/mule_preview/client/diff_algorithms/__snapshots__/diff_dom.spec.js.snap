// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when diffing two MAST data structures produces correct output 1`] = `
Array [
  Object {
    "action": "removeElement",
    "element": Object {
      "attributes": Object {
        "description": "example:/expunge-convolvers",
        "hash": 1066677628,
        "tag-name": "sub-flow",
      },
      "childNodes": Array [
        Object {
          "attributes": Object {
            "description": "Flush the convolver prisms",
            "hash": 579843109,
            "tag-name": "db:no-operation-selected",
          },
          "childNodes": null,
          "nodeName": "db:no-operation-selected",
          "original-attributes": Array [],
          "original-type": "component",
        },
        Object {
          "attributes": Object {
            "description": "Did the flush occur?",
            "hash": -1105159260,
            "tag-name": "choice",
          },
          "childNodes": Array [
            Object {
              "attributes": Object {
                "description": null,
                "hash": 2014296490,
                "tag-name": "when",
              },
              "childNodes": Array [
                Object {
                  "attributes": Object {
                    "description": "Fetch more prisms",
                    "hash": 881435866,
                    "tag-name": "s3:no-operation-selected",
                  },
                  "childNodes": null,
                  "nodeName": "s3:no-operation-selected",
                  "original-attributes": Array [],
                  "original-type": "component",
                },
                Object {
                  "attributes": Object {
                    "description": "Log number of fanglehockeys",
                    "hash": 543925965,
                    "tag-name": "logger",
                  },
                  "childNodes": null,
                  "nodeName": "logger",
                  "original-attributes": Array [],
                  "original-type": "component",
                },
              ],
              "nodeName": "when",
              "original-attributes": Array [
                "horizontal",
              ],
              "original-type": "container",
            },
            Object {
              "attributes": Object {
                "description": null,
                "hash": -15128758,
                "tag-name": "otherwise",
              },
              "childNodes": Array [
                Object {
                  "attributes": Object {
                    "description": "Log Error",
                    "hash": -818462707,
                    "tag-name": "logger",
                  },
                  "childNodes": null,
                  "nodeName": "logger",
                  "original-attributes": Array [],
                  "original-type": "component",
                },
              ],
              "nodeName": "otherwise",
              "original-attributes": Array [
                "horizontal",
              ],
              "original-type": "container",
            },
          ],
          "nodeName": "choice",
          "original-attributes": Array [
            "vertical",
          ],
          "original-type": "container",
        },
      ],
      "nodeName": "sub-flow",
      "original-attributes": Array [
        "horizontal",
      ],
      "original-type": "container",
    },
    "route": Array [
      1,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Object {
        "description": "Overwrite the payload",
        "hash": 266487046,
        "tag-name": "set-payload",
      },
      "childNodes": null,
      "nodeName": "set-payload",
      "original-attributes": Array [],
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      0,
    ],
  },
  Object {
    "action": "removeElement",
    "element": Object {
      "attributes": Object {
        "description": "Filter recalcitrant ions",
        "hash": -2096291590,
        "tag-name": "expression-filter",
      },
      "childNodes": null,
      "nodeName": "expression-filter",
      "original-attributes": Array [],
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Object {
        "description": "Nothing to see here",
        "hash": 54737936,
        "tag-name": "set-payload",
      },
      "childNodes": null,
      "nodeName": "set-payload",
      "original-attributes": Array [],
      "original-type": "component",
    },
    "route": Array [
      0,
      1,
      0,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "description",
    "newValue": "Set Z level to 7000",
    "oldValue": "Set Z level to 9000",
    "route": Array [
      0,
      0,
      2,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "hash",
    "newValue": 641237277,
    "oldValue": 1907014272,
    "route": Array [
      0,
      0,
      2,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Object {
        "description": "example:/strain-overflow-noodle-requests",
        "hash": -2070961420,
        "tag-name": "flow-ref",
      },
      "childNodes": null,
      "nodeName": "flow-ref",
      "original-attributes": Array [],
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
      0,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Object {
        "description": "example:/strain-overflow-noodle-requests",
        "hash": -2070961420,
        "tag-name": "flow-ref",
      },
      "childNodes": null,
      "nodeName": "flow-ref",
      "original-attributes": Array [],
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
      1,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Object {
        "description": "example:/strain-overflow-noodle-requests",
        "hash": -2070961420,
        "tag-name": "flow-ref",
      },
      "childNodes": null,
      "nodeName": "flow-ref",
      "original-attributes": Array [],
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
      2,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Object {
        "description": "example:/strain-overflow-noodle-requests",
        "hash": -2070961420,
        "tag-name": "flow-ref",
      },
      "childNodes": null,
      "nodeName": "flow-ref",
      "original-attributes": Array [],
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
      3,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "hash",
    "newValue": 251600214,
    "oldValue": 385704797,
    "route": Array [
      0,
      0,
      7,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Object {
        "description": "Server Noodles 2",
        "hash": 798826535,
        "tag-name": "set-payload",
      },
      "childNodes": null,
      "nodeName": "set-payload",
      "original-attributes": Array [],
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      8,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "description",
    "newValue": "Overwrite the payload",
    "oldValue": "Nothing to see here",
    "route": Array [
      0,
      1,
      0,
      1,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "hash",
    "newValue": 266487046,
    "oldValue": 54737936,
    "route": Array [
      0,
      1,
      0,
      1,
    ],
  },
]
`;
